Add Organization DTOs and unit tests

marierinden opened the pull request:
### Description

This pull request introduces Data Transfer Objects (DTOs) and a resource for managing organizations, along with unit tests to validate their functionality.

**Changes Included:**
- Added `OrganizationResponse` and `OrganizationCollectionResponse` DTOs for representing organization data.
- Added `OrganizationRequest` DTO for handling input data.
- Implemented `OrganizationResource` for interfacing with the organizations API.
- Created unit tests for `OrganizationRequest`, `OrganizationResponse`, and `OrganizationCollectionResponse` DTOs to ensure proper structure, data handling, and behavior.

marierinden committed:
Add DTOs and resource for Organization management

Introduce `OrganizationResponse` and `OrganizationCollectionResponse` DTOs to represent organization data. Add `OrganizationRequest` for input handling, and `OrganizationResource` for interfacing with the organizations API.

marierinden committed:
Add unit tests for Organization DTO request and response models

Introduced tests for `OrganizationRequest`, `OrganizationResponse`, and `OrganizationCollectionResponse` classes. These tests validate the structure, data handling, and expected behaviors of the DTO models, ensuring reliability and correctness.

fredrik-tveraaen-apiera commented on src/DTO/Request/Organization/OrganizationRequest.php:
```diff
@@ -0,0 +1,66 @@
+     */
+    public function __construct(
+        #[RequestField('name')]
+        private string $name,
```
Change to nullable

fredrik-tveraaen-apiera commented on src/DTO/Request/Organization/OrganizationRequest.php:
```diff
@@ -0,0 +1,66 @@
+        #[RequestField('name')]
+        private string $name,
+        #[RequestField('extId')]
+        private string $extId,
```
Change to nullable

fredrik-tveraaen-apiera commented on src/DTO/Request/Organization/OrganizationRequest.php:
```diff
@@ -0,0 +1,66 @@
+        private string $name,
+        #[RequestField('extId')]
+        private string $extId,
+        #[RequestField('alternateIdentifiers')]
+        private array $alternateIdentifiers = [],
```
Remove this as it's an documentation bug

fredrik-tveraaen-apiera commented on src/DTO/Request/Organization/OrganizationRequest.php:
```diff
@@ -0,0 +1,66 @@
+    ) {
+    }
+
+    public function getName(): string
```
Change to nullable

fredrik-tveraaen-apiera commented on src/DTO/Request/Organization/OrganizationRequest.php:
```diff
@@ -0,0 +1,66 @@
+        return $this->name;
+    }
+
+    public function getExtId(): string
```
Change to nullable

fredrik-tveraaen-apiera commented on src/DTO/Response/Organization/OrganizationResponse.php:
```diff
@@ -0,0 +1,95 @@
+        #[ResponseField('extId')]
+        private string $extId,
+        #[ResponseField('alternateIdentifiers')]
+        private array $alternateIdentifiers = [],
```
Remove this as it's an documentation bug

fredrik-tveraaen-apiera commented on src/Resource/OrganizationResource.php:
```diff
@@ -0,0 +1,166 @@
+ */
+final readonly class OrganizationResource implements RequestResourceInterface
+{
+    private const string ENDPOINT = '/organizations';
```
Add prefix /api/v1/ as organizations is not a nested endpoint

fredrik-tveraaen-apiera commented on src/Resource/OrganizationResource.php:
```diff
@@ -0,0 +1,166 @@
+        $collection = $this->find($request, $params);
+
+        if ($collection->getLdTotalItems() < 1) {
+            throw new InvalidRequestException('No Organization found matching the given criteria');
```
Change Organization to lowercase

marierinden committed:
Update OrganizationRequest and Resource classes

Make `name` and `extId` nullable in `OrganizationRequest` and remove `alternateIdentifiers` field. Adjust endpoint URL in `OrganizationResource` and improve exception message formatting for consistency.

marierinden committed:
Refactor OrganizationResponse to remove alternateIdentifiers

Removed the `alternateIdentifiers` property and its related methods from `OrganizationResponse`. Updated `name` and `extId` fields to be nullable. Adjusted corresponding test cases to align with these changes.

fredrik-tveraaen-apiera commented on tests/Unit/DTO/Response/Organization/OrganizationCollectionResponseTest.php:
```diff
@@ -0,0 +1,57 @@
+    protected function getCollectionData(): array
+    {
+        $organizationResponse = new OrganizationResponse(
+            ldId: '/api/v1/stores/123/products/123',
```
Change to /api/v1/organizations/123

fredrik-tveraaen-apiera commented on tests/Unit/DTO/Response/Organization/OrganizationCollectionResponseTest.php:
```diff
@@ -0,0 +1,57 @@
+    {
+        $organizationResponse = new OrganizationResponse(
+            ldId: '/api/v1/stores/123/products/123',
+            ldType: LdType::Product,
```
Change to LdType::Organization

fredrik-tveraaen-apiera commented on tests/Unit/DTO/Response/Organization/OrganizationCollectionResponseTest.php:
```diff
@@ -0,0 +1,57 @@
+            uuid: Uuid::fromString('bfd2639c-7793-426a-a413-ea262e582208'),
+            createdAt: new DateTimeImmutable('2021-01-01 00:00:00'),
+            updatedAt: new DateTimeImmutable('2021-01-01 00:00:00'),
+            name: 'Product',
```
Change to a more readable example (e.g. string, Organization test)

fredrik-tveraaen-apiera commented on tests/Unit/DTO/Response/Organization/OrganizationCollectionResponseTest.php:
```diff
@@ -0,0 +1,57 @@
+        );
+
+        return [
+            'ldContext' => '/api/v1/contexts/Product',
```
Change to '/api/v1/contexts/Organization'

fredrik-tveraaen-apiera commented on tests/Unit/DTO/Response/Organization/OrganizationCollectionResponseTest.php:
```diff
@@ -0,0 +1,57 @@
+
+        return [
+            'ldContext' => '/api/v1/contexts/Product',
+            'ldId' => '/api/v1/stores/123/products',
```
Change to /api/v1/organizations

fredrik-tveraaen-apiera commented on tests/Unit/DTO/Response/Organization/OrganizationCollectionResponseTest.php:
```diff
@@ -0,0 +1,57 @@
+            'ldMembers' => [$organizationResponse],
+            'ldTotalItems' => 1,
+            'ldView' => new PartialCollectionView(
+                ldId: '/api/v1/stores/123/products?page=1',
+                ldFirst: '/api/v1/stores/123/products?page=1',
+                ldLast: '/api/v1/stores/123/products?page=1',
```
Change to /api/v1/organizations?page=1

fredrik-tveraaen-apiera commented on tests/Unit/DTO/Response/Organization/OrganizationResponseTest.php:
```diff
@@ -0,0 +1,51 @@
+    protected function getResponseData(): array
+    {
+        return [
+            'ldId' => '/api/v1/stores/123/products/123',
```
/api/v1/organizations/123

fredrik-tveraaen-apiera commented on tests/Unit/DTO/Response/Organization/OrganizationResponseTest.php:
```diff
@@ -0,0 +1,51 @@
+            'uuid' => Uuid::fromString('bfd2639c-7793-426a-a413-ea262e582208'),
+            'createdAt' => new DateTimeImmutable('2021-01-01 00:00:00'),
+            'updatedAt' => new DateTimeImmutable('2021-01-01 00:00:00'),
+            'name' => 'Product',
```
Change to a more readable example (e.g. string, Organization test)

fredrik-tveraaen-apiera commented on tests/Unit/DTO/Response/Organization/OrganizationResponseTest.php:
```diff
@@ -0,0 +1,51 @@
+    {
+        return [
+            'ldId' => '/api/v1/stores/123/products/123',
+            'ldType' => LdType::Product,
```
Change to LdType::Organization

fredrik-tveraaen-apiera commented on tests/Unit/DTO/Response/Organization/OrganizationResponseTest.php:
```diff
@@ -0,0 +1,51 @@
+
+    protected function getExpectedLdType(): LdType
+    {
+        return LdType::Product;
```
Change to LdType::Organization

marierinden committed:
Update tests to reflect Organization-related data changes

Replaced references to "Product" with "Organization" in test data and expectations. Adjusted URLs, names, and types to align with the new focus on organizations for consistency and correctness in test cases.

fredrik-tveraaen-apiera committed:
Add integration tests and support for Organization resource

Introduced CRUD operations for the Organization resource, including Create, Read, Update, and Delete tests. Updated the SDK to include OrganizationResource, its response types, and handling in LdType. These changes extend functionality and ensure proper integration testing.

fredrik-tveraaen-apiera committed:
Update test data URL in OrganizationResponseTest

fredrik-tveraaen-apiera committed:
Make 'name' and 'extId' fields non-nullable

Updated the `OrganizationResponse` DTO to ensure `name` and `extId` are required fields by removing their nullability. Adjusted related tests to match these changes by modifying the `getNullableFields` method.

fredrik-tveraaen-apiera reviewed (approved):
:100:
